### PR TITLE
Storage Lockdown -1 - Refactor subnets related components 

### PR DESCRIPTION
**Pull request by mjudeikis**
### Which issue this PR addresses:

No new features, just a refactor in preparation to merge and to storage lockdown.

### What this PR does / why we need it:

This is needed so we can re-use some of the `util/subnets` functions. It addition it makes subnets NSG bit cleaner. 

### Test plan for issue:

Unit tests, e2e

### Is there any documentation that needs to be updated for this PR?

No

### CODE REVIEW

File: pkg/util/subnet/subnet.go [hunk 1]
```diff
@@ -13,25 +13,43 @@ import (
...
+}
+
 type Manager interface {
 	Get(ctx context.Context, subnetID string) (*mgmtnetwork.Subnet, error)
 	GetHighestFreeIP(ctx context.Context, subnetID string) (string, error)
 	CreateOrUpdate(ctx context.Context, subnetID string, subnet *mgmtnetwork.Subnet) error
+
+	// ListFromCluster uses kube clients, instead of azure go get cluster
+	// subnets
+	ListFromCluster(ctx context.Context) ([]Subnet, error)
```
Comment (m1kola) [inline pkg/util/subnet/subnet.go:35]: Somehow this doesn't seem like a good fit: now this interface both deals with Azure API and with clsuters... Responsibility seem to be too broad. Maybe we need a separate thing for interacting with clusters?

Also I don't like the fact that in many cases we are going to call `NewManager` with `nil` as a first argument which is going to be confusing: you have to know when to provide it and when not to provide it.

What do you think?

**Review by m1kola — COMMENTED** [review]
Still reviewing, but wanted to share some early thoughts on `ListFromCluster`.

File: pkg/util/subnet/subnet.go (hunk 1 above)
Comment (mjudeikis) [reply pkg/util/subnet/subnet.go:35]: I dont like it too, but we need it in a re-usable way. So open for suggestions. Separate interface? Separate package?

Comment (m1kola) [reply pkg/util/subnet/subnet.go:35]: @mjudeikis Yes, I was think about a separate interface & struct for it. I think same package is ok.

File: pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go [hunk 2]
```diff
@@ -45,25 +45,26 @@ type InternetCheckerSpec struct {
...
+	APIIntIP               string              `json:"apiIntIP,omitempty"`
+	IngressIP              string              `json:"ingressIP,omitempty"`
 
 	Features FeaturesSpec `json:"features,omitempty"`
 }
 
 // FeaturesSpec defines ARO operator feature gates
 type FeaturesSpec struct {
-	ReconcileNSGs                  bool `json:"reconcileNSGs,omitempty"`
+	ReconcileSubnets               bool `json:"ReconcileSubnets,omitempty"`
```
Comment (jewzaam) [inline pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go:67]:
```suggestion
	ReconcileSubnets               bool `json:"reconcileSubnets,omitempty"`
```

File: pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go [hunk 3]
```diff
@@ -45,25 +45,26 @@ type InternetCheckerSpec struct {
...
+	VnetID                 string              `json:"vnetId,omitempty"`
+	APIIntIP               string              `json:"apiIntIP,omitempty"`
+	IngressIP              string              `json:"ingressIP,omitempty"`
 
 	Features FeaturesSpec `json:"features,omitempty"`
 }
 
 // FeaturesSpec defines ARO operator feature gates
 type FeaturesSpec struct {
-	ReconcileNSGs                  bool `json:"reconcileNSGs,omitempty"`
```
Comment (jewzaam) [inline pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go:66]: Recommend we keep `reconcileNSGs` and deprecate unless we know for sure that it's not used anywhere.  The PR could include a migration of values from `reconcileNSGs` to `reconcileSubnets` but without this behavior in clusters will be impacted.

**Review by jewzaam — CHANGES_REQUESTED** [review]
Initial review, hope to come back around for another look.  Generally looks reasonable except want to be sure we don't enable the controller on clusters where it's disabled with `reconcileNSGs`.

File: pkg/util/subnet/cluster_subnets_test.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+			expectedErr: fmt.Errorf("unexpected end of JSON input"),
+		},
+		{
+			name:        "missing master nodes",
+			expectedMap: nil,
+			modify: func(worker *machinev1beta1.Machine, master *machinev1beta1.Machine) {
+				master.Labels = map[string]string{}
+			},
+			expectedErr: fmt.Errorf("master resource group not found"),
+		},
```
[the remark this answers is not in the document]
Comment (m1kola) [reply pkg/util/subnet/cluster_subnets_test.go:62]: We usually do something like this:

```go
if err != nil && err.Error() != tt.wantErr ||
	err == nil && tt.wantErr != "" {
	t.Error(err)
}
```

So we the test fails wehen we expect an eror, but there is no error.

File: pkg/operator/controllers/subnets/subnetnsg_test.go [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+				if err != nil {
+					return nil, err
+				}
+
+				machine2, err := getMachineObject("worker", vnetResourceGroup, vnetName, subnetNameWorker, false)
+				if err != nil {
+					return nil, err
+				}
+
+				return maofake.NewSimpleClientset(machine1, machine2), nil
```
Comment (m1kola) [inline pkg/operator/controllers/subnets/subnetnsg_test.go:114]: I would just panic in `getMachineObject` when there is an error: it is a developer error so it is ok I think. With panic we can do something like:

```go
maocli: maofake.NewSimpleClientset(
	getMachineObject("master", vnetResourceGroup, vnetName, subnetNameMaster, true),
	getMachineObject("worker", vnetResourceGroup, vnetName, subnetNameWorker, false),
)
```

Comment (m1kola) [reply pkg/operator/controllers/subnets/subnetnsg_test.go:114]: Also, `maocli` function currenlty seems to be exactly the same in all tests cases. Does it make sense to move this into the test body?

**Review by m1kola — CHANGES_REQUESTED** [review]

File: pkg/operator/controllers/const.go [hunk 6]
```diff
@@ -5,7 +5,7 @@ package controllers
 
 const (
 	AlertwebhookControllerName             = "Alertwebhook"
-	AzureNSGControllerName                 = "AzureNSG"
+	AzureSubnetsControllerName             = "AzureNSG"
```
Comment (nilsanderselde) [inline pkg/operator/controllers/const.go:8]: Does the value of this constant need to remain `AzureNSG`?

File: pkg/operator/controllers/subnets/subnetnsg_test.go (hunk 5 above)
Comment (mjudeikis) [reply pkg/operator/controllers/subnets/subnetnsg_test.go:114]: I dont think it is even needed now when kubeSubnet interface was split...

File: pkg/operator/controllers/const.go (hunk 6 above)
Comment (mjudeikis) [reply pkg/operator/controllers/const.go:8]: no :)

File: pkg/util/subnet/cluster_subnets_test.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+					t.Errorf("Expected Error %s, got %s when processing %s testcase", tt.expectedErr.Error(), err.Error(), tt.name)
+				}
+				return
+			}
+
+			if tt.expectedMap != nil {
+				if len(tt.expectedMap) != len(subnets) {
+					t.Errorf("Expected Map length %d, doesn't match result map length %d when processing %s testcase", len(tt.expectedMap), len(subnets), tt.name)
+				}
+				if cmp.Equal(tt.expectedErr, subnets) {
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply pkg/util/subnet/cluster_subnets_test.go]: copy paste while refactoring original code...

**Review by m1kola — CHANGES_REQUESTED** [review]
Bumping an old question about migration from previous values to a new + `pkg/util/subnet/cluster_subnets_test.go` seems to be broken (it is passing, but not doing what it is intended to do).

Everything else is just minor suggestions.

File: pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go (hunk 3 above)
Comment (mjudeikis) [reply pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go:75]: @Makdaam how we are dealing with this controller now in prod?

Comment (mjudeikis) [reply pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go:75]: Confirmed with @Makdaam , we only few enabled and there was some fallback. So its safe to do `false` as we gonna revisit this in prod.

File: pkg/operator/controllers/subnets/subnetnsg_test.go [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+			r := reconcileManager{
+				log:            log,
+				instance:       instance,
+				subscriptionID: subscriptionId,
+				subnets:        subnets,
+				kubeSubnets:    kubeSubnets,
+			}
+
+			err := r.reconcileSubnets(context.Background())
+			if err != nil {
```
[the remark this answers is not in the document]
Comment (m1kola) [reply pkg/operator/controllers/subnets/subnetnsg_test.go:248]: Also: I do not see where we use `wantErr` at all. If we have error in the return value - we should test it, right? So we are probably not covering some cases.

**Review by m1kola — APPROVED** [review]
One last thing in tests. I'm happy for it to be addressed in a separate PR, if you prefer.

**Pull request merged**